File: client/communication/TCPProtocol.go
package communication

import (
	"encoding/binary"
	"net"
)

type TCPProtocol struct {
	conn net.Conn
}

func NewTCPProtocol(conn net.Conn) TCPProtocol {
	return TCPProtocol{conn: conn}
}

// SendAll sends the given data over the TCP connection.
func (tcpProt TCPProtocol) SendAll(data []byte) error {
	totalSend := 0
	for totalSend < len(data) {
		n, err := tcpProt.conn.Write(data[totalSend:])
		if err != nil {
			return err
		}
		totalSend += n
	}
	return nil
}

// RecvAll receives data from the TCP connection. It first reads
// 2 bytes to determine the length of the incoming message, and
// then reads the message data based on that length.
// It returns the complete message, including the length bytes.
func (tcpProt TCPProtocol) RecvAll() ([]byte, error) {
	lengthBytes, err := tcpProt.RecvExact(2)
	if err != nil {
		return nil, err
	}
	length := binary.BigEndian.Uint16(lengthBytes)

	data, err := tcpProt.RecvExact(int(length))
	if err != nil {
		return nil, err
	}

	return append(lengthBytes, data...), nil
}

// RecvExact reads exactly n bytes from the TCP connection.
func (tcpProt TCPProtocol) RecvExact(n int) ([]byte, error) {
	data := make([]byte, n)
	totalRead := 0
	for totalRead < n {
		bytesRead, err := tcpProt.conn.Read(data[totalRead:])
		if err != nil {
			return nil, err
		}
		totalRead += bytesRead
	}
	return data, nil
}

func (tcpProt TCPProtocol) Close() {
	tcpProt.conn.Close()
}

File: server/communication/tcp_protocol.py
import socket


class TCPProtocol:
    def __init__(self):
        pass

    def send_all(self, conn: socket.socket, data: bytes):
        """
        Sends all bytes to the TCP connection.
        """
        conn.sendall(data)

    def recv_with_length(self, conn: socket.socket):
        """
        Receives all bytes from the TCP connection, first reading the length of the data 
        (2 bytes) and then reading the data itself.
        """
        length_bytes = self.recv_exact(conn, 2)
        length = int.from_bytes(length_bytes, byteorder='big', signed=False)

        data = self.recv_exact(conn, length)

        return length_bytes + data

    def recv_exact(self, conn: socket.socket, n: int) -> bytes:
        """
        Receives exactly n bytes from the TCP connection. If the connection is closed 
        before n bytes are received, raises a ConnectionError.
        """
        data = b''
        while len(data) < n:
            chunk = conn.recv(n - len(data))
            if not chunk:
                raise ConnectionError("connection closed")
            data += chunk
        return data